lib/generate-critical-css.js

```diff
@@ -1,26 +1,84 @@
 const CSSFileSet = require( './css-file-set' );
 const { removeIgnoredPseudoElements } = require( './ignored-pseudo-elements' );
 const { minifyCss } = require( './minify-css' );
+const { SuccessTargetError } = require( './errors' );
+const { BrowserInterface } = require( './browser-interface' );
+
+/**
+ * Collate and return a CSSFileSet object describing all the CSS files used by
+ * the set of URLs provided.
+ *
+ * URLs which fail to load generate warnings, but do not stop the process.
+ *
+ * @param {BrowserInterface} browserInterface - interface to access pages
+ * @param {string[]} urls - list of URLs to scan for CSS files
+ * @param {number} successUrlsThreshold - success urls amount threshold
+ * @return {CSSFileSet} - Set of CSS files used by the urls.
+ */
+async function collateCssFiles( browserInterface, urls, successUrlsThreshold ) {
+	const cssFiles = new CSSFileSet();
+	const warnings = [];
+	let successes = 0;
+
+	for ( const url of urls ) {
+		try {
+			const cssIncludes = await browserInterface.getCssIncludes( url );
+
+			// Convert relative URLs to absolute.
+			const relativeUrls = Object.keys( cssIncludes );
+			const absoluteIncludes = relativeUrls.reduce( ( set, relative ) => {
+				const absolute = new URL( relative, url ).toString();
+				set[ absolute ] = cssIncludes[ relative ];
+
+				return set;
+			}, {} );
+
+			await cssFiles.addMultiple( url, absoluteIncludes );
+
+			// Abort early if we hit the threshold of success urls.
```

Comment:
Now that we're using a success threshold instead of a min/max, I no longer think that `collateCssFiles` should abort early once the minimum success threshold has been met. There are some critical problems that might arise here:

1. If the first URL is successfully read, then the success rate will be 100% at this stage, and it will happily abort early. Then, no other URLs will be examined.
2. If a URL works at this stage, but fails at a later stage, it will have been removed from the valid URL set - and the overall process will fail, even if another URL in the set might be valid.

IMHO, now that we're using success thresholds we should simplify and count successes in the main processing loop - then evaluate if the threshold has been met at the end.

We can't abort early just based on a success threshold %, unless we _also_ have a minimum absolute number of URLs that must be successfully processed too. Otherwise we'll stop after the first sucess, and that would be bad. :)

Reply:
Oh, I retract the above statements. I realized after posting this that you are calculating an absolute number of urls that must succeed based on the number supplied, and so my comments about bailing after one are off the mark!